src/fsharp/NameResolution.fs

```diff
@@ -3962,21 +3979,31 @@ let rec ResolvePartialLongIdentInModuleOrNamespaceForRecordFields (ncenv: NameRe
            |> List.filter (fun tycon -> tycon.IsRecordTycon)
            |> List.filter (fun tycon -> not (IsTyconUnseen ad g ncenv.amap m (modref.NestedTyconRef tycon)))
 
-       let ilTyconNames = 
-           mty.TypesByAccessNames.Values
-           |> List.choose (fun (tycon:Tycon) -> if tycon.IsILTycon then Some tycon.DisplayName else None)
-           |> Set.ofList
+
+       let accessibleSubModules =
+           let moduleOrNamespaces =
+               mty.ModulesAndNamespacesByDemangledName 
+               |> NameMap.range
+
+           if isNil moduleOrNamespaces then [] else
+
+           let ilTyconNames =
+               let hashSet = HashSet()
+               for tycon in mty.TypesByAccessNames.Values do
+                   if tycon.IsILTycon then 
+                       hashSet.Add tycon.DisplayName |> ignore
+               hashSet
     
-        // Collect up the accessible sub-modules 
-       (mty.ModulesAndNamespacesByDemangledName 
-          |> NameMap.range 
-          |> List.filter (fun x -> 
-                let demangledName = x.DemangledModuleOrNamespaceName
-                notFakeContainerModule ilTyconNames demangledName && IsInterestingModuleName demangledName)
-          |> List.map modref.NestedTyconRef
-          |> List.filter (IsTyconUnseen ad g ncenv.amap m >> not)
-          |> List.filter (EntityRefContainsSomethingAccessible ncenv m ad)
-          |> List.map ItemForModuleOrNamespaceRef)
+           moduleOrNamespaces
```

Comment:
It looks to me there is some duplicate code here, even before this PR. Perhaps in a separate PR we should try to make a common function contain this.